README.md
# db-docker
Generate a Database as an image

## Introduction

This tool was written for Axelerant specific processes but can be used generally by overriding certain configuration. By default, you wouldn't be able to use this plugin if you cannot access Axelerant's GitLab repository. You may use it with your own Docker image registry or even with Docker Hub.

## Prerequisites

* Reasonably updated version of composer with recent version of PHP. Tested with composer 1.10.1 and composer 2+.
* Recent version of Docker.

### Optional requirements (only for default workflow)

By default, this plugin generates the image with name prefixed with Axelerant's Docker registry. The below requirements apply if you don't override the image name using configuration in [composer.json](#configuration).

* Ability to clone from gitorious.xyz.
* Logged in to Axelerant's GitLab Container Registry. To verify, run `docker login registry.gitorious.xyz`. Optional if you use the `--no-push` option.

## Installation

Install with composer into any Drupal project.

```bash
composer require --dev axelerant/db-docker
```

## Usage

Options may be specified using the command line or by specifying them in the `extra` section in your `composer.json` file. See the section below on [Configuration](#configuration) for more details.

```
$ composer db-docker --help
Usage:
  db-docker [options]

Options:
  -t, --docker-tag[=DOCKER-TAG]  The Docker tag to build
  -r, --git-remote[=GIT-REMOTE]  The git remote to use to determine the image name
  -s, --db-source[=DB-SOURCE]    Source of the database ("lando", "drush", or "file")
  -f, --db-file[=DB-FILE]        The path to the database file (required if db-source is set to file)
      --no-push                  Set to not push the image after building
  -h, --help                     Display this help message
  -q, --quiet                    Do not output any message
  -V, --version                  Display this application version
      --ansi                     Force ANSI output
      --no-ansi                  Disable ANSI output
  -n, --no-interaction           Do not ask any interactive question
      --profile                  Display timing and memory usage information
      --no-plugins               Whether to disable plugins.
  -d, --working-dir=WORKING-DIR  If specified, use the given directory as working directory.
      --no-cache                 Prevent use of the cache
  -v|vv|vvv, --verbose           Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Help:
  Generate a Docker image for the database.
```

**Warning:** The above usage is generated by running `composer db-docker --help` and is generated by composer. Some options such as `--no-plugins` are not compatible. The option `--no-plugins` tells composer not to load any plugins, which means that `db-docker` won't be loaded at all.

### Examples

To let the plugin guess defaults, build the image, and push it.

```bash
composer db-docker
```

To explicitly specify a SQL file to build the image.

```bash
composer db-docker --db-source=file --db-file=<filename> # The file can either be plain SQL or gzipped.
```

## Configuration

The plugin also supports configuration via composer.json `extra` section.

```json
{
    "name": "drupal/site",
    // ...
    "require": {
        // ...
    },
    "require-dev": {
        // ...
        "axelerant/db-docker": "^1.0"
    },
    "extra": {
        "dbdocker": {
            "docker-image-name": "auto",
            "docker-tag": "auto",
            "docker-base": {
                "base-flavor": "bitnami",
                "image": "bitnami/mariadb:10.4",
                "user": "drupal8",
                "password": "drupal8",
                "database": "drupal8"
            },
            "git-remote": "origin",
            "db-source": "",
            "no-push": false
        }
    }
}
```

These options work the same way as the options available via the command line. Option values specified on the command line take precedence followed by the options in the `extra` section of `composer.json`. Finally, values are guessed as described below.

### Image details

By default, the image name and tag will be guessed using the method described below in the [Default Options](#default-options) section. You can override both of them using the `docker-image-name` and `docker-tag` settings in the configuration in `composer.json` as shown above. Additionally, you may also specify the `docker-tag` via the command line.

### Base image

You can use the `docker-base` configuration to specify the base image to use to build the database image. When not specified, the default image used is Bitnami's MariaDB 10.4 image with default access details which is compatible with [Lando's Drupal 8 recipe](https://docs.lando.dev/config/drupal8.html). You may override this to fit in your workflow as desired.

The base image details cannot be customized from the command line. They must be specified via the `composer.json`'s extra section.

### DDEV base image

The defaults for `docker-base` depend on the value of `base-flavor` setting. By default, the `base-flavor` is set to `bitnami` and the defaults are set as above. However, if `base-flavor` is `ddev`, the defaults change to the following:

```json
{
    "extra": {
        "dbdocker": {
            "docker-base": {
                "base-flavor": "ddev",
                "image": "drud/ddev-dbserver-mariadb-10.4:v1.17.0",
                "user": "db",
                "password": "db",
                "database": "db"
            }
        }
    }
}
```

These are the defaults required for building database images compatible for DDEV.

## Default options

The plugin tries to guess most values for input to correctly select the source, build the image, and push it.

### Determining the image name

The image name can be specified using the `docker-image-name` in `composer.json` as described above. If not specified (or set to `"auto"`), the image name is constructed from the git repository.

The image name is determined based on the git repository's `origin` remote (overridable using the `--git-remote` option). The remote URL should be a Git URL (not a HTTP URL) of type `git@gitorious.xyz:<group>/<project>.git`. For this, it would determine the image name `registry.gitorious.xyz/<group>/<project>/db`. See the next section for the image tag.

It's important to note that the image name is constructed only for Axelerant Gitlab repositories. If you want to use another Docker registry (such as Docker Hub or Quay.io), please use the option in `composer.json` to specify the proper name.

### Determining the image tag

The image tag, unless specified with the `--docker-tag` option, is assumed to be the current branch name. If the current branch is `master`, the image tag is used as `latest`.

### Determining the database source

These database sources are supported: `file`, `lando`, `ddev`, and `drush`. The source can be explicitly specified using the `--db-source` option. If not specified, the following rules are used to determine the source.
* If the `--db-file` option is present, then the source is set as `file`.
* If a file called `.lando.yml` is present, then the source is set as `lando`.
    * As an exception to above, the plugin attempts to detect if it is running inside a lando container. If so, the source is set to `drush`.
* If a directory called `.ddev` is present, then the source is set as `ddev`.
    * As an exception to above, the plugin attempts to detect if it is running inside a DDEV container. If so, the source is set to `drush`.
* If the above conditions fail, then the source is assumed to be `drush`.

In case the source is `lando`, `ddev`, or `drush`, the `drush sql:dump` command is used to obtain the SQL file. If the source is `lando` or `ddev`, then the drush command is executed inside the relevant container like so: `lando drush ...` or `ddev drush ...`.

## Reporting problems

If you see a bug or an improvement, create a pull request. For support, raise a request on Axelerant Slack #internal-support channel.


CHANGELOG.md
# Changelog

## UNRELEASED

* ...

## v1.1.3

* Add support for MariaDB 10.6 for DDEV
* Fix #25: Return an exit code
* Assume tag is 'latest' if the current branch is 'main'

## v1.1.2

* List ddev as a db-source by @skippednote in #21
* Add support for GitHub URLs by @skippednote in #23

## v1.1.1

* Set db-source to file when db-file option is present. Fixes #17.

## v1.1.0

* Update documentation for DDEV support
* Fix #15: Add support for DDEV
* Remove an invalid SQL statement in truncate caches file
* Fix the ARG order in Dockerfile

## v1.0.1

* Add support for symfony/process 2.8. Fixes #12

## v1.0.0

* Set a high timeout value so that docker build and push have enough time
* Add documentation for docker-base section

## v1.0.0-beta10

* Warn if the image is not something we recognize. Fixes #11
* Add support for specifying a base image. Fixes #11
* Add a partial package config test

## v1.0.0-beta9

* Fix issue with the flipped behaviour of --no-push parameter

## v1.0.0-beta8

* Add support for any Docker registry. Fixes #10
* Merge pull request #9 from axelerant/options-extra
* Fix hash parameters for GH Actions caching
* Add phpunit to Github Actions
* Add tests for OptionsProvider
* Document how to configure the plugin using the `extra` section
* Support configuration from composer.json extra section. Fixes #8
* Clarify the usage instructions in README
* Fixes #5: Do not query remotes when reading information
* Merge pull request #2 from axelerant/phplint
* Add php lint and cs workflow
* Add phplint
* Use symfony/process for running commands
* Reformat code to take lesser space
* Fix comment in DbDocker
* Enforce PSR-12 standards
* Update CHANGELOG for beta7 release

## v1.0.0-beta7

* Correctly determine paths when building images

## v1.0.0-beta6

* Minor improvements
* Remove a TODO after testing
* Move the dockerize-db Dockerfile and SQL inside the plugin

## v1.0.0-beta5

* Remove shortcut for no-push

## v1.0.0-beta4

* Add support for composer 2.0
* Merge pull request #1 from axelerant/arguments
* Update single hyphen arugment name


tests/OptionsProviderTest.php
<?php

namespace Axelerant\DbDocker\Tests;

use Axelerant\DbDocker\OptionsProvider;
use Composer\Package\RootPackageInterface;
use PHPUnit\Framework\TestCase;
use Prophecy\Prophet;
use Symfony\Component\Console\Input\InputInterface;

class OptionsProviderTest extends TestCase
{
    protected $prophet;

    protected $input;

    protected $package;

    public function setUp(): void
    {
        parent::setUp();
        $this->prophet = new Prophet();
        $this->input = $this->prophet->prophesize(InputInterface::class);
        $this->package = $this->prophet->prophesize(RootPackageInterface::class);
    }

    /**
     * @dataProvider dataOptions
     */
    public function testRegularOptions($input, $package, $expected): void
    {
        $this->input->getOption('docker-tag')->willReturn($input['docker-tag']);
        $this->input->getOption('git-remote')->willReturn($input['git-remote']);
        $this->input->getOption('db-source')->willReturn($input['db-source']);
        $this->input->getOption('db-file')->willReturn($input['db-file']);
        $this->input->getOption('no-push')->willReturn($input['no-push']);
        $this->package->getExtra()->willReturn(['dbdocker' => $package]);

        $options = new OptionsProvider($this->input->reveal(), $this->package->reveal());

        $this->assertSame($expected['docker-image-name'], $options->getDockerImageName());
        $this->assertSame($expected['docker-tag'], $options->getDockerTag());
        $this->assertSame($expected['git-remote'], $options->getGitRemote());
        $this->assertSame($expected['db-source'], $options->getDbSource());
        $this->assertSame($expected['push'], $options->getPush());

        $base = $options->getDockerBaseDetails();
        $this->assertSame($expected['docker-base']['image'], $base['image']);
        $this->assertSame($expected['docker-base']['user'], $base['user']);
        $this->assertSame($expected['docker-base']['password'], $base['password']);
        $this->assertSame($expected['docker-base']['database'], $base['database']);
    }

    public function dataOptions(): array
    {
        $defaultInput = [
            'docker-tag' => '',
            'git-remote' => '',
            'db-source' => '',
            'db-file' => '',
            'no-push' => false,
        ];
        $defaultExpected = [
            'docker-image-name' => 'auto',
            'docker-tag' => 'auto',
            'docker-base' => [
                'base-flavor' => 'bitnami',
                'image' => 'bitnami/mariadb:10.4',
                'user' => 'drupal8',
                'password' => 'drupal8',
                'database' => 'drupal8',
            ],
            'git-remote' => 'origin',
            'db-source' => '',
            'push' => true,
        ];

        $cases = [];

        // All defaults
        $cases[] = [
            'input' => $defaultInput,
            'package' => [],
            'expected' => $defaultExpected,
        ];

        // Configuration in package extra.
        $cases[] = [
            'input' => $defaultInput,
            'package' => [
                'docker-image-name' => 'test/image',
                'docker-tag' => 'latest',
                'git-remote' => 'upstream',
                'db-source' => 'lando',
                'no-push' => false,
            ],
            'expected' => [
                'docker-image-name' => 'test/image',
                'docker-tag' => 'latest',
                'git-remote' => 'upstream',
                'db-source' => 'lando',
                'push' => true,
            ] + $defaultExpected,
        ];

        // Configuration overridden on CLI.
        $cases[] = [
            'input' => [
                'docker-tag' => 'production',
                'git-remote' => 'upstream',
                'db-source' => '',
                'no-push' => true,
            ] + $defaultInput,
            'package' => [
                'docker-image-name' => 'auto',
                'docker-tag' => 'latest',
                'git-remote' => '',
                'db-source' => 'lando',
                'no-push' => false,
            ],
            'expected' => [
                'docker-image-name' => 'auto',
                'docker-tag' => 'production',
                'git-remote' => 'upstream',
                'db-source' => 'lando',
                'push' => false,
            ] + $defaultExpected,
        ];

        // Partial configuration in composer.json.
        $cases[] = [
            'input' => [
                'docker-tag' => 'production',
                'git-remote' => 'upstream',
                'db-source' => '',
                'no-push' => true,
            ] + $defaultInput,
            'package' => [
                'docker-tag' => 'latest',
                'db-source' => 'lando',
            ],
            'expected' => [
                'docker-image-name' => 'auto',
                'docker-tag' => 'production',
                'git-remote' => 'upstream',
                'db-source' => 'lando',
                'push' => false,
            ] + $defaultExpected,
        ];

        // Input of db-file should always set the db-source option to 'file'.
        // See https://github.com/axelerant/db-docker/issues/17.
        $cases[] = [
            'input' => [
                'db-source' => '',
                'db-file' => 'file.sql',
            ] + $defaultInput,
            'package' => [
                'db-source' => 'lando',
            ],
            'expected' => [
                'db-source' => 'file',
            ] + $defaultExpected,
        ];

        // Configuration with docker-base.
        $cases[] = [
            'input' => $defaultInput,
            'package' => [
                'docker-base' => [
                    'image' => 'mariadb:latest',
                    'user' => 'mysql',
                    'password' => 'root',
                    'database' => 'drupal',
                ],
            ],
            'expected' => [
                'docker-base' => [
                    'image' => 'mariadb:latest',
                    'user' => 'mysql',
                    'password' => 'root',
                    'database' => 'drupal',
                ],
            ] + $defaultExpected,
        ];

        // Partial configuration with docker-base.
        $cases[] = [
            'input' => $defaultInput,
            'package' => [
                'docker-tag' => 'latest',
                'docker-base' => [
                    'image' => 'mariadb:latest',
                ],
                'db-source' => 'lando',
            ],
            'expected' => [
                    'docker-tag' => 'latest',
                    'docker-base' => [
                        'image' => 'mariadb:latest',
                        'user' => 'drupal8',
                        'password' => 'drupal8',
                        'database' => 'drupal8',
                    ],
                    'db-source' => 'lando',
                ] + $defaultExpected,
        ];

        // Partial configuration with DDEV base flavor.
        $cases[] = [
            'input' => $defaultInput,
            'package' => [
                'docker-tag' => 'latest',
                'docker-base' => [
                    'base-flavor' => 'ddev',
                ],
                'db-source' => 'lando',
            ],
            'expected' => [
                    'docker-tag' => 'latest',
                    'docker-base' => [
                        'base-flavor' => 'ddev',
                        'image' => 'drud/ddev-dbserver-mariadb-10.4:v1.17.0',
                        'user' => 'db',
                        'password' => 'db',
                        'database' => 'db',
                    ],
                    'db-source' => 'lando',
                ] + $defaultExpected,
        ];

        return $cases;
    }
}


src/OptionsProvider.php
<?php

namespace Axelerant\DbDocker;

use Composer\Package\RootPackageInterface;
use Symfony\Component\Console\Input\InputInterface;

/**
 * Class OptionsProvider
 * @package Axelerant\DbDocker
 *
 * Get options first from the command line input and if not available, from the
 * configuration in composer.json file.
 */
class OptionsProvider
{
    /**
     * @var InputInterface
     */
    protected $input;
    /**
     * @var array
     */
    protected $packageConfig;

    public function __construct(InputInterface $input, RootPackageInterface $package)
    {
        $this->input = $input;
        $extra = $package->getExtra();
        $this->packageConfig = $extra['dbdocker'] ?? [];
        $this->packageConfig += [
            'docker-image-name' => 'auto',
            'docker-tag' => 'auto',
            'docker-base' => [],
            'git-remote' => '',
            'db-source' => '',
            'no-push' => false,
        ];

        $baseImageDetailsDefault = [
            'base-flavor' => 'bitnami',
            'image' => 'bitnami/mariadb:10.4',
            'user' => 'drupal8',
            'password' => 'drupal8',
            'database' => 'drupal8',
        ];
        $baseFlavor = $this->packageConfig['docker-base']['base-flavor'] ?? 'default';
        if ($baseFlavor == 'ddev') {
            $baseImageDetailsDefault = [
                'image' => 'drud/ddev-dbserver-mariadb-10.4:v1.17.0',
                'user' => 'db',
                'password' => 'db',
                'database' => 'db',
            ];
        }
        $this->packageConfig['docker-base'] += $baseImageDetailsDefault;
    }

    public function getDockerTag(): string
    {
        return $this->input->getOption('docker-tag') ?: $this->packageConfig['docker-tag'];
    }

    public function getDockerImageName(): string
    {
        return $this->packageConfig['docker-image-name'];
    }

    public function getDockerBaseDetails(): array
    {
        return $this->packageConfig['docker-base'];
    }

    public function getGitRemote(): string
    {
        return $this->input->getOption('git-remote') ?: $this->packageConfig['git-remote'] ?: 'origin';
    }

    public function getDbSource(): string
    {
        $dbSource = $this->input->getOption('db-source') ?: $this->packageConfig['db-source'];
        // If the db-file option is set, always set db-source to file.
        // See https://github.com/axelerant/db-docker/issues/17.
        if ($this->input->getOption('db-file')) {
            $dbSource = 'file';
        }

        return $dbSource;
    }

    public function getPush(): bool
    {
        return !($this->input->getOption('no-push') ?: $this->packageConfig['no-push']);
    }
}


src/DbDockerCommand.php
<?php

namespace Axelerant\DbDocker;

use Composer\Command\BaseCommand;
use GitElephant\Repository;
use Symfony\Component\Console\Exception\InvalidOptionException;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Process\Exception\ProcessFailedException;
use Symfony\Component\Process\Process;
use Symfony\Component\Process\ProcessBuilder;

class DbDockerCommand extends BaseCommand
{
    /**
     * @var InputInterface
     */
    protected $input;

    /**
     * @var OutputInterface
     */
    protected $output;

    /**
     * @var OptionsProvider
     */
    protected $options;

    /**
     * {@inheritDoc}
     */
    protected function configure()
    {
        $this->setName('db-docker')
            ->setDescription('Generate a Docker image for the database.')
            ->addOption(
                'docker-tag',
                't',
                InputOption::VALUE_OPTIONAL,
                'The Docker tag to build'
            )
            ->addOption(
                'git-remote',
                'r',
                InputOption::VALUE_OPTIONAL,
                'The git remote to use to determine the image name'
            )
            ->addOption(
                'db-source',
                's',
                InputOption::VALUE_OPTIONAL,
                'Source of the database ("lando", "ddev", "drush", or "file")'
            )
            ->addOption(
                'db-file',
                'f',
                InputOption::VALUE_OPTIONAL,
                'The path to the database file (required if db-source is set to file)'
            )
            ->addOption(
                'no-push',
                null,
                InputOption::VALUE_NONE,
                'Set to not push the image after building'
            );
    }

    /**
     * {@inheritDoc}
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $this->input = $input;
        $this->output = $output;
        $this->options = new OptionsProvider($input, $this->getComposer()->getPackage());

        $imageId = $this->getImageId();

        $sqlFile = $this->getDbFile();

        $baseDetails = $this->options->getDockerBaseDetails();
        $this->verifyBaseImage($baseDetails);
        $this->buildImage($imageId, $sqlFile, $baseDetails);

        if ($this->options->getPush()) {
            $this->output->writeln("<info>Pushing image...</info>");
            $this->execCmd(['docker', 'push', $imageId]);
        } else {
            $this->output->writeln(sprintf("<info>To push image, run '%s'</info>", "docker push " . $imageId));
        }

        return 0;
    }

    /**
     * Get the complete image name (with tag) based on given options.
     */
    protected function getImageId(): string
    {
        // We can safely use `getcwd()` even in a subdirectory.
        $git = new Repository(getcwd());
        $tag = $this->options->getDockerTag();
        if (!$tag || $tag == 'auto') {
            $tag = $git->getMainBranch()->getName();
            $this->output->writeln("<info>Docker tag not specified. Using current branch name: {$tag}</info>");

            // We should be using the tag 'latest' if the current branch is 'master' or 'main'.
            if ($tag == 'master' || $tag == 'main') {
                $tag = 'latest';
                $this->output->writeln(
                    "<info>Using Docker tag 'latest' for branch '{$tag}'.</info>",
                    OutputInterface::VERBOSITY_VERBOSE
                );
            }
        }

        $imageName = $this->options->getDockerImageName();
        if (!$imageName || $imageName == "auto") {
            // Throws an exception if the remote not found, so we don't have to.
            $remote = $git->getRemote($this->options->getGitRemote(), false);
            $imageName = $this->getImageNameFromRepoUrl($remote->getFetchURL());
            $this->output->writeln("<info>Docker image not specified. Using from git repository: {$imageName}</info>");
        }

        // Determine the image name (path) from the git remote URL.
        return sprintf("%s:%s", $imageName, $tag);
    }

    /**
     * Determine the Docker image name from the repo URL.
     *
     * @param string $url
     *
     * @return string
     */
    protected function getImageNameFromRepoUrl(string $url): string
    {
        if (!preg_match('/^[^@]*@([^:]*):(.*)\.git$/', $url, $matches)) {
            throw new InvalidOptionException("The specified git remote URL couldn't be parsed");
        }

        $host = $matches[1];
        $path = $matches[2];
        switch ($host) {
            case 'gitlab.axl8.xyz':
                $registryDomain = 'registry.axl8.xyz';
                break;
            case 'gitorious.xyz':
            case 'code.axelerant.com':
                $registryDomain = 'registry.gitorious.xyz';
                break;
            case 'github.com':
                $registryDomain = 'ghcr.io';
                break;
            default:
                throw new InvalidOptionException("The specified git remote URL isn't supported");
        }

        return sprintf("%s/%s/db", $registryDomain, strtolower($path));
    }

    /**
     * Get the database file path based on the source.
     */
    protected function getDbFile(): string
    {
        $src = $this->options->getDbSource() ?: $this->guessSource();
        $validOptions = ['lando', 'ddev', 'drush', 'file'];
        if (!in_array($src, $validOptions)) {
            throw new InvalidOptionException("db-source can only be one of 'lando', 'ddev', 'drush', or 'file'");
        }

        $this->output->writeln("<info>Getting SQL file from source '{$src}'</info>");

        if ($src == 'file') {
            if (!$this->input->getOption('db-file')) {
                throw new InvalidOptionException("db-file is required if db-source is set to 'file'");
            }
            return realpath($this->input->getOption('db-file'));
        }

        // Get SQL from Lando or Drush.
        $sqlFileName = tempnam(sys_get_temp_dir(), 'axldb');
        $drushCmd = 'drush sql:dump > ' . $sqlFileName;
        if ($src == 'lando') {
            $drushCmd = 'lando ' . $drushCmd;
        }
        if ($src == 'ddev') {
            $drushCmd = 'ddev ' . $drushCmd;
        }

        $this->execCmd($drushCmd);
        return $sqlFileName;
    }

    /**
     * Guess the best database source.
     */
    protected function guessSource(): string
    {
        // If there is a file called '.lando.yml', there is a good chance
        // that the project uses lando and we should use that for the source.
        if (file_exists('.lando.yml')) {
            // If we are running inside Lando, just use 'drush'.
            return getenv('LANDO') == 'ON' ? 'drush' : 'lando';
        }
        // Similarly, if there is a directory called `.ddev`, we are probably
        // running ddev.
        if (file_exists('.ddev')) {
            // If we are running inside DDEV, just use 'drush'.
            return getenv('IS_DDEV_PROJECT') ? 'drush' : 'ddev';
        }

        return 'drush';
    }

    /**
     * Show warnings if the image doesn't appear to be a MariaDB image.
     *
     * @param array $baseImageDetails
     */
    protected function verifyBaseImage(array $baseImageDetails): void
    {
        $image = $baseImageDetails['image'];
        if (!preg_match("/^([^:]+)(:.+)?$/", $image, $matches)) {
            throw new \UnexpectedValueException("Cannot parse Docker image name: '{$image}.'");
        }

        $imageName = $matches[1];
        $allowedImages = [
            'bitnami/mariadb',
            'mariadb',
            'drud/ddev-dbserver-mariadb-10.2',
            'drud/ddev-dbserver-mariadb-10.3',
            'drud/ddev-dbserver-mariadb-10.4',
            'drud/ddev-dbserver-mariadb-10.5',
            'drud/ddev-dbserver-mariadb-10.6',
        ];
        if (!in_array(strtolower($imageName), $allowedImages)) {
            $this->output->writeln(
                "<comment>Cannot recognize image name '{$imageName}'. Use at your own risk.</comment>"
            );
        }
    }

    /**
     * Build the image using our Dockerfile and SQL scripts
     *
     * @param string $imageId
     * @param string $sqlFile
     * @param array $baseDetails
     */
    protected function buildImage(string $imageId, string $sqlFile, array $baseDetails): void
    {
        $tempDir = realpath(sys_get_temp_dir());
        $tempPath = sprintf('%s%s%s', $tempDir, DIRECTORY_SEPARATOR, sha1(uniqid())) . '/';

        $dockerfileName = $baseDetails['base-flavor'] == 'ddev' ? 'dockerize-db-ddev' : 'dockerize-db';
        $assetPath = realpath(__DIR__ . '/../assets/' . $dockerfileName) . '/';

        mkdir($tempPath);
        mkdir($tempPath . 'dumps');
        copy($assetPath . 'Dockerfile', $tempPath . 'Dockerfile');
        copy($sqlFile, $tempPath . "/dumps/db.sql");
        copy($assetPath . "zzzz-truncate-caches.sql", $tempPath . "zzzz-truncate-caches.sql");

        // Little hacky but this works for now
        if (file_exists($assetPath . 'create_init_db.sh')) {
            copy($assetPath . 'create_init_db.sh', $tempPath . 'create_init_db.sh');
            chmod($tempPath . 'create_init_db.sh', 0755);
        }

        $this->output->writeln("<info>Building image '{$imageId}'</info>");
        $dockerCmd = ['docker', 'build', '-t', $imageId];
        $dockerCmd[] = '--build-arg';
        $dockerCmd[] = sprintf("BASE_IMAGE=%s", $baseDetails['image']);
        $dockerCmd[] = '--build-arg';
        $dockerCmd[] = sprintf("BASE_IMAGE_USER=%s", $baseDetails['user']);
        $dockerCmd[] = '--build-arg';
        $dockerCmd[] = sprintf("BASE_IMAGE_PASSWORD=%s", $baseDetails['password']);
        $dockerCmd[] = '--build-arg';
        $dockerCmd[] = sprintf("BASE_IMAGE_DATABASE=%s", $baseDetails['database']);
        $dockerCmd[] = $tempPath;
        $this->execCmd($dockerCmd);
    }

    protected function execCmd($cmd): void
    {
        $this->output->writeln(sprintf(
            "<info>Running '%s'</info>",
            is_array($cmd) ? implode(" ", $cmd) : $cmd
        ), OutputInterface::VERBOSITY_VERBOSE);

        if (is_array($cmd) && PHP_VERSION_ID < 70400 && class_exists(ProcessBuilder::class)) {
            // Composer embeds an old version of symfony/process, and we need to
            // target that. This version (2.8.52 as of this writing) does not
            // support command line as an array which results in an error with
            // proc_open before PHP 7.4. This is the case we should target.
            // See https://github.com/axelerant/db-docker/issues/12.
            $p = (new ProcessBuilder($cmd))->getProcess();
        } elseif (is_string($cmd) && method_exists(Process::class, 'fromShellCommandline')) {
            // BC for symfony/process < 4.2.
            // The method fromShellCommandline is new in 4.2 and it deprecated
            // using strings for constructor (and was removed in 5). Since we are
            // trying to support a variety of versions, this check is necessary when
            // the command is a string.
            // This code is only left here in case composer updates its version of
            // symfony/process one day. In most cases, this line of code will not
            // be executed when using composer phar file.
            $p = Process::fromShellCommandline($cmd);
        } else {
            // Versions of symfony/process from 3.3 to before 5.0 supported
            // constructor with string AND array parameters.
            $p = new Process($cmd);
        }

        $p->setTimeout(600);
        $code = $p->run();
        $this->output->writeln($p->getOutput(), OutputInterface::OUTPUT_RAW | OutputInterface::VERBOSITY_VERBOSE);

        if (!$p->isSuccessful()) {
            $this->output->writeln("<error>Command returned exit code '{$code}'</error>");
            throw new ProcessFailedException($p);
        }
    }
}
